Sources/SourceKitLSP/Swift/SwiftLanguageServer.swift

```diff
@@ -978,143 +963,202 @@ extension SwiftLanguageServer {
-              currentComment!.length += length
-              return true
-            }
-            if let comment = currentComment {
-              self.addFoldingRange(offset: comment.offset, length: comment.length, kind: .comment, in: snapshot, toSet: &ranges)
-            }
-            currentComment = (offset: offset, length: length)
-          }
-          return hasReachedLimit == false
+        override func visit(_ node: ClosureExprSyntax) -> SyntaxVisitorContinueKind {
+          return self.addFoldingRange(
+            start: node.statements.position.utf8Offset,
+            end: node.rightBrace.positionAfterSkippingLeadingTrivia.utf8Offset)
         }
 
-        // Add the last stored comment.
-        if let comment = currentComment, hasReachedLimit == false {
-          self.addFoldingRange(offset: comment.offset, length: comment.length, kind: .comment, in: snapshot, toSet: &ranges)
-          currentComment = nil
+        override func visit(_ node: AccessorBlockSyntax) -> SyntaxVisitorContinueKind {
+          return self.addFoldingRange(
+            start: node.accessors.position.utf8Offset,
+            end: node.rightBrace.positionAfterSkippingLeadingTrivia.utf8Offset)
         }
 
-        var structureStack: [SKDResponseArray] = [substructure]
-        while !hasReachedLimit, let substructure = structureStack.popLast() {
-          substructure.forEach { _, value in
-            if let offset: Int = value[self.keys.bodyoffset],
-               let length: Int = value[self.keys.bodylength],
-               length > 0
-            {
-              self.addFoldingRange(offset: offset, length: length, in: snapshot, toSet: &ranges)
-              if hasReachedLimit {
-                return false
-              }
-            }
-            if let substructure: SKDResponseArray = value[self.keys.substructure] {
-              structureStack.append(substructure)
+        override func visit(_ node: SwitchStmtSyntax) -> SyntaxVisitorContinueKind {
+          return self.addFoldingRange(
+            start: node.cases.position.utf8Offset,
+            end: node.rightBrace.positionAfterSkippingLeadingTrivia.utf8Offset)
+        }
+
+        override func visit(_ node: FunctionCallExprSyntax) -> SyntaxVisitorContinueKind {
+          return self.addFoldingRange(
+            start: node.argumentList.position.utf8Offset,
+            end: node.argumentList.endPosition.utf8Offset)
+        }
+
+        override func visit(_ node: SubscriptExprSyntax) -> SyntaxVisitorContinueKind {
+          return self.addFoldingRange(
+            start: node.argumentList.position.utf8Offset,
+            end: node.argumentList.endPosition.utf8Offset)
+        }
+
+        __consuming func finalize() -> Set<FoldingRange> {
+          return self.ranges
+        }
+
+        private func addFoldingRange(start: Int, end: Int, kind: FoldingRangeKind? = nil) -> SyntaxVisitorContinueKind {
```

Comment:
Is this returning a `SyntaxVisitorContinueKind` so you can `return self.addFoldingRange` in the `visit` methods? To me it implied that this method had some smart logic whether it should visit children or not and then I was surprised that it always returned `visitChildren`.

I think it would be good if you could either
- Add a doc comment like: ``This method always returns `.visitChildren` so `visit` method implementations can be `return self.addFoldingRange(…)` `` or
- Have this method return `Void`

Reply:
This was always meant to return `.skipChildren` if the number of ranges exceeded the requested limit. Thanks.